File: app/src/main/java/com/jhj/jshare/jshare/JShareUtil.java
package com.jhj.jshare.jshare;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.AsyncTask;
import android.text.TextUtils;
import android.util.Log;
import android.widget.Toast;

import com.jhj.jshare.MainActivity;
import com.jhj.jshare.MyApplication;
import com.jhj.jshare.R;
import com.jhj.jshare.jshare.bean.ImgShareBuilder;
import com.jhj.jshare.jshare.bean.LinkShareBuilder;
import com.jhj.jshare.jshare.bean.TextShareBuilder;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.util.HashMap;

import javax.net.ssl.HttpsURLConnection;

import cn.jiguang.share.android.api.JShareInterface;
import cn.jiguang.share.android.api.PlatActionListener;
import cn.jiguang.share.android.api.Platform;
import cn.jiguang.share.android.api.ShareParams;
import cn.jiguang.share.qqmodel.QQ;
import cn.jiguang.share.qqmodel.QZone;
import cn.jiguang.share.wechat.Wechat;
import cn.jiguang.share.wechat.WechatFavorite;
import cn.jiguang.share.wechat.WechatMoments;
import cn.jiguang.share.weibo.SinaWeibo;

/**
 * @author吕志豪 .
 * @date 18-4-14  上午9:39.
 * Github :https://github.com/lvzhihao100
 * E-Mail：1030753080@qq.com
 * 简书 :http://www.jianshu.com/u/6e525b929aac
 */

public class JShareUtil {

    private ShareParams shareParams;
    private ImgShareBuilder imgShareBuilder;
    private TextShareBuilder textShareBuilder;
    private LinkShareBuilder linkShareBuilder;
    private ShareType type;
    private Activity baseActivity;


    public JShareUtil(ImgShareBuilder imgShareBuilder, Activity baseActivity) {
        this.baseActivity = baseActivity;
        this.imgShareBuilder = imgShareBuilder;
        type = ShareType.IMG;
    }

    public JShareUtil(TextShareBuilder textShareBuilder, Activity baseActivity) {
        this.baseActivity = baseActivity;
        this.textShareBuilder = textShareBuilder;
        type = ShareType.TEXT;
    }

    public JShareUtil(LinkShareBuilder linkShareBuilder, Activity baseActivity) {
        this.baseActivity = baseActivity;
        this.linkShareBuilder = linkShareBuilder;
        type = ShareType.LINK;
    }

    public void share2EQD() {
        //todo
        switch (type) {
            case IMG:
                Intent intent = new Intent(Intent.ACTION_SEND);
                intent.setType("image/*");
                intent.putExtra(Intent.EXTRA_STREAM, Uri.fromFile(new File(imgShareBuilder.getImagePath())));
                intent.setComponent(new ComponentName("com.eqdd.yiqidian", "com.eqdd.yiqidian.ui.share.ShareEnterActivity"));
                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                baseActivity.startActivity(intent);
                break;
            case TEXT:
                intent = new Intent(Intent.ACTION_SEND);
                intent.setType("text/plain"); // 纯文本
                intent.putExtra(Intent.EXTRA_SUBJECT, textShareBuilder.getTitle());
                intent.putExtra(Intent.EXTRA_TEXT, textShareBuilder.getText());
                intent.setComponent(new ComponentName("com.eqdd.yiqidian", "com.eqdd.yiqidian.ui.share.ShareEnterActivity"));
                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                baseActivity.startActivity(intent);
                break;
            case LINK:
              /*  LinkContent linkContent = new LinkContent(linkShareBuilder.getTitle(), linkShareBuilder.getText(), linkShareBuilder.getUrl(),
                        linkShareBuilder.getImageUrl(), linkShareBuilder.getSource(), linkShareBuilder.getId());
                AppConstent.messageContent = LinkMessage.obtain(linkContent);
                baseActivity.startActivity(new Intent(baseActivity, ShareEnterActivity.class));
                break;*/
        }
    }

    public void share2WorkCircle() {
        switch (type) {
            case IMG:

                break;
            case TEXT:

                break;
            case LINK:
                //todo 根据需求写
                Intent intent = new Intent(baseActivity, MainActivity.class);
                intent.putExtra("title", linkShareBuilder.getTitle());
                intent.putExtra("image", linkShareBuilder.getImageUrl());
                intent.putExtra("source", linkShareBuilder.getSource());
                intent.putExtra("url", linkShareBuilder.getUrl());
                intent.putExtra("id", linkShareBuilder.getId());
                baseActivity.startActivity(intent);
                break;
        }
    }

    public void share2Collection() {
        //todo 根据需求写,如果是添加到收藏,大概是将参数提交到网络请求
        switch (type) {
            case IMG:
                CollectionUtil.INSTANCE.chooseImg(imgShareBuilder.getImageUrl(), imgShareBuilder.getSource(), imgShareBuilder.getSourceOwner());
                break;
            case TEXT:
                CollectionUtil.INSTANCE.chooseText(textShareBuilder.getTitle(), textShareBuilder.getText(), textShareBuilder.getSource(), textShareBuilder.getSourceOwner());
                break;
            case LINK:
                CollectionUtil.INSTANCE.chooseLink(linkShareBuilder.getTitle(), linkShareBuilder.getUrl(), linkShareBuilder.getSource(),
                        linkShareBuilder.getSourceOwner(), linkShareBuilder.getId());
                break;

        }
    }

    public void share2WeXin(String platform) {
        switch (type) {
            case IMG:
                shareParams = new ShareParams();
                shareParams.setShareType(Platform.SHARE_IMAGE);
                if (!TextUtils.isEmpty(imgShareBuilder.getImagePath())) {
                    shareParams.setImagePath(imgShareBuilder.getImagePath());
                    share(platform, shareParams);
                } else {
                    DownImage image = new DownImage(platform, shareParams);
                    image.execute(imgShareBuilder.getImageUrl());
                }
                break;
            case TEXT:
                shareParams = new ShareParams();
                shareParams.setShareType(Platform.SHARE_TEXT);
                shareParams.setText(textShareBuilder.getText());//必须
                share(platform, shareParams);
                break;
            case LINK:
                shareParams = new ShareParams();
                shareParams.setTitle(linkShareBuilder.getTitle());
                shareParams.setText(linkShareBuilder.getText());
                shareParams.setShareType(Platform.SHARE_WEBPAGE);
                shareParams.setUrl(linkShareBuilder.getUrl());//必须
                DownImage image = new DownImage(platform, shareParams);
                image.execute(linkShareBuilder.getImageUrl());
                break;

        }


    }

    public void share2Q(String platform) {
        switch (type) {
            case IMG:
                shareParams = new ShareParams();
                shareParams.setShareType(Platform.SHARE_IMAGE);
                if (TextUtils.isEmpty(imgShareBuilder.getImageUrl())) {
                    shareParams.setImagePath(imgShareBuilder.getImagePath());
                } else {
                    shareParams.setImageUrl(imgShareBuilder.getImageUrl());
                }
                break;
            case TEXT:
                shareParams = new ShareParams();
                shareParams.setShareType(Platform.SHARE_TEXT);
                shareParams.setText(textShareBuilder.getText());
                shareParams.setTitle(textShareBuilder.getTitle());
                break;
            case LINK:
                shareParams = new ShareParams();
                shareParams.setTitle(linkShareBuilder.getTitle());
                shareParams.setText(linkShareBuilder.getText());
                shareParams.setShareType(Platform.SHARE_WEBPAGE);
                shareParams.setUrl(linkShareBuilder.getUrl());//必须
                if (TextUtils.isEmpty(linkShareBuilder.getImageUrl())) {
                    shareParams.setImagePath(linkShareBuilder.getImagePath());
                } else {
                    shareParams.setImageUrl(linkShareBuilder.getImageUrl());
                }
                break;
        }
        share(platform, shareParams);
    }

    public void share2Sina() {
        switch (type) {
            case TEXT:
                shareParams = new ShareParams();
                shareParams.setShareType(Platform.SHARE_TEXT);
                shareParams.setText(textShareBuilder.getText());
                shareParams.setTitle(textShareBuilder.getTitle());
                share(SinaWeibo.Name, shareParams);
                break;
            case IMG:
                shareParams = new ShareParams();
                shareParams.setShareType(Platform.SHARE_IMAGE);
                if (!TextUtils.isEmpty(imgShareBuilder.getImagePath())) {
                    shareParams.setImagePath(imgShareBuilder.getImagePath());
                    share(SinaWeibo.Name, shareParams);
                } else {
                    DownImage image = new DownImage(SinaWeibo.Name, shareParams);
                    image.execute(imgShareBuilder.getImageUrl());
                }

                break;
            case LINK:
                shareParams = new ShareParams();
                shareParams.setShareType(Platform.SHARE_WEBPAGE);
                shareParams.setText(linkShareBuilder.getText());
                shareParams.setUrl(linkShareBuilder.getUrl());
                DownImage image = new DownImage(SinaWeibo.Name, shareParams);
                image.execute(linkShareBuilder.getImageUrl());
        }
    }

    public void share2Explorer() {
        Intent in = new Intent(Intent.ACTION_VIEW, Uri.parse(linkShareBuilder.getUrl()));
        baseActivity.startActivity(in);
    }


    enum ShareType {
        TEXT,
        IMG,
        LINK,
        FILE,
        VIDEO
    }

    static void share(String platForm, ShareParams shareParams) {

        if (platForm.equals(QQ.Name)) {
            if (shareParams.getText() != null && shareParams.getTitle().length() > 30) {
                shareParams.setTitle(shareParams.getTitle().substring(0, 30));
            }

            if (shareParams.getText() != null && shareParams.getText().length() > 40) {
                shareParams.setText(shareParams.getText().substring(0, 40));
            }
        }


        JShareInterface.share(platForm, shareParams, new PlatActionListener() {
            @Override
            public void onComplete(Platform platform, int i, HashMap<String, Object> hashMap) {
                System.out.println("分享成功");
            }

            @Override
            public void onError(Platform platform, int i, int i1, Throwable throwable) {
                System.out.println("分享失败" + throwable.toString());
                if (platform.getName().equals(QQ.Name) || platform.getName().equals(QZone.Name)) {
                    Toast.makeText(MyApplication.instance, "请安装QQ", Toast.LENGTH_SHORT).show();
                } else if (platform.getName().equals(Wechat.Name) || platform.getName().equals(WechatFavorite.Name)
                        || platform.getName().equals(WechatMoments.Name)) {
                    Toast.makeText(MyApplication.instance, "请安装Wechat", Toast.LENGTH_SHORT).show();

                } else if (platform.getName().equals(SinaWeibo.Name)) {
                    Toast.makeText(MyApplication.instance, "请安装微博", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onCancel(Platform platform, int i) {
                System.out.println("分享取消");
            }
        });
    }


    /**
     * 分享时,将图片链接转为bitmap
     */
    static class DownImage extends AsyncTask<String, Bitmap, Bitmap> {

        private String platform;
        private ShareParams shareParams;

        DownImage(String platform, ShareParams shareParams) {
            this.platform = platform;
            this.shareParams = shareParams;
        }

        @Override
        protected Bitmap doInBackground(String... params) {
            Bitmap bitmap = null;
            try {
                String strUrl = params[0];
                URL url = new URL(strUrl);
                HttpsURLConnection conn = (HttpsURLConnection) url.openConnection();
                conn.setRequestMethod("GET");
                conn.setConnectTimeout(5 * 1000);
                InputStream inStream = conn.getInputStream();//通过输入流获取图片数据
                bitmap = BitmapFactory.decodeStream(inStream);
            } catch (Exception e) {
                bitmap = BitmapFactory.decodeResource(MyApplication.instance.getResources(), R.mipmap.ic_launcher);
            }
            return bitmap;
        }

        @Override
        protected void onPostExecute(Bitmap bitmap) {
            //bitmap = compressBitmap(bitmap);
            shareParams.setImageData(bitmap);
            share(platform, shareParams);
        }

        /**
         * 当需要限制bitmap大小时
         * 将图片压缩到指定大小之下
         *
         * @param bitmap 原bitmap
         * @return 目标bitmap
         * @throws IOException
         */
        private Bitmap compressBitmap(Bitmap bitmap) throws IOException {
            int topLimit = 30 * 1024;
            if (bitmap != null) {
                ByteArrayOutputStream baos = new ByteArrayOutputStream();
                int options = 100;
                bitmap.compress(Bitmap.CompressFormat.JPEG, options, baos);
                Log.w("xxx", baos.toByteArray().length + "");
                while (baos.toByteArray().length > topLimit) {
                    baos.reset();
                    options -= 5;
                    bitmap.compress(Bitmap.CompressFormat.JPEG, options, baos);
                }
                ByteArrayInputStream isBm = new ByteArrayInputStream(baos.toByteArray());// 把压缩后的数据baos存放到ByteArrayInputStream中
                bitmap = BitmapFactory.decodeStream(isBm, null, null);// 把ByteArrayInputStream数据生成
                isBm.close();
            }
            return bitmap;
        }

    }


}

File: app/src/main/java/com/jhj/jshare/jshare/JShareUtil.kt
package com.jhj.jshare.jshare

import android.app.Activity
import android.content.ComponentName
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.text.TextUtils
import android.util.Log
import cn.jiguang.share.android.api.JShareInterface
import cn.jiguang.share.android.api.PlatActionListener
import cn.jiguang.share.android.api.Platform
import cn.jiguang.share.android.api.ShareParams
import cn.jiguang.share.qqmodel.QQ
import cn.jiguang.share.qqmodel.QZone
import cn.jiguang.share.wechat.Wechat
import cn.jiguang.share.wechat.WechatFavorite
import cn.jiguang.share.wechat.WechatMoments
import cn.jiguang.share.weibo.SinaWeibo
import com.jhj.jshare.MainActivity
import com.jhj.jshare.MyApplication
import com.jhj.jshare.R
import com.jhj.jshare.jshare.bean.ImgShareBuilder
import com.jhj.jshare.jshare.bean.LinkShareBuilder
import com.jhj.jshare.jshare.bean.TextShareBuilder
import org.jetbrains.anko.doAsyncResult
import org.jetbrains.anko.toast
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException
import java.net.URL
import java.util.*
import javax.net.ssl.HttpsURLConnection

/**
 *
 * Created by jhj on 18-5-25.
 */

class JShareUtil() {

    private lateinit var imgShareBuilder: ImgShareBuilder
    private lateinit var textShareBuilder: TextShareBuilder
    private lateinit var linkShareBuilder: LinkShareBuilder
    private var type: ShareType? = null
    private var mActivity: Activity? = null


    enum class ShareType {
        TEXT,
        IMG,
        LINK,
        /*  FILE,
          VIDEO*/
    }


    constructor(imgShareBuilder: ImgShareBuilder, mActivity: Activity) : this() {
        this.mActivity = mActivity
        this.imgShareBuilder = imgShareBuilder
        type = ShareType.IMG
    }

    constructor(textShareBuilder: TextShareBuilder, mActivity: Activity) : this() {
        this.mActivity = mActivity
        this.textShareBuilder = textShareBuilder
        type = ShareType.TEXT
    }

    constructor(linkShareBuilder: LinkShareBuilder, mActivity: Activity) : this() {
        this.mActivity = mActivity
        this.linkShareBuilder = linkShareBuilder
        type = ShareType.LINK
    }

    fun share2EQD() {
        //todo
        when (type) {
            JShareUtil.ShareType.IMG -> {
                val intent = Intent(Intent.ACTION_SEND)
                intent.type = "image/*"
                intent.putExtra(Intent.EXTRA_STREAM, Uri.fromFile(File(imgShareBuilder.imagePath)))
                intent.component = ComponentName("com.eqdd.yiqidian", "com.eqdd.yiqidian.ui.share.ShareEnterActivity")
                intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                mActivity?.startActivity(intent)
            }
            JShareUtil.ShareType.TEXT -> {
                val intent = Intent(Intent.ACTION_SEND)
                intent.setType("text/plain") // 纯文本
                intent.putExtra(Intent.EXTRA_SUBJECT, textShareBuilder.title)
                intent.putExtra(Intent.EXTRA_TEXT, textShareBuilder.text)
                intent.setComponent(ComponentName("com.eqdd.yiqidian", "com.eqdd.yiqidian.ui.share.ShareEnterActivity"))
                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                mActivity?.startActivity(intent)
            }
        }
    }

    fun share2WorkCircle() {
        when (type) {
            JShareUtil.ShareType.IMG -> {
            }
            JShareUtil.ShareType.TEXT -> {
            }
            JShareUtil.ShareType.LINK -> {
                //todo 根据需求写
                val intent = Intent(mActivity, MainActivity::class.java)
                intent.putExtra("title", linkShareBuilder.title)
                intent.putExtra("image", linkShareBuilder.imageUrl)
                intent.putExtra("source", linkShareBuilder.source)
                intent.putExtra("url", linkShareBuilder.url)
                intent.putExtra("id", linkShareBuilder.id)
                mActivity?.startActivity(intent)
            }
        }
    }

    fun share2Collection() {
        //todo 根据需求写,如果是添加到收藏,大概是将参数提交到网络请求
        when (type) {
            JShareUtil.ShareType.IMG -> CollectionUtil.chooseImg(imgShareBuilder.imageUrl, imgShareBuilder.source, imgShareBuilder.sourceOwner)
            JShareUtil.ShareType.TEXT -> CollectionUtil.chooseText(textShareBuilder.title, textShareBuilder.text, textShareBuilder.source, textShareBuilder.sourceOwner)
            JShareUtil.ShareType.LINK -> CollectionUtil.chooseLink(linkShareBuilder.title, linkShareBuilder.url, linkShareBuilder.source,
                    linkShareBuilder.sourceOwner, linkShareBuilder.id)
        }
    }

    fun share2WeXin(platform: String) {
        val shareParams = ShareParams()
        when (type) {
            JShareUtil.ShareType.TEXT -> {
                shareParams.shareType = Platform.SHARE_TEXT
                shareParams.text = textShareBuilder.text//必须
                share(platform, shareParams)
            }
            JShareUtil.ShareType.IMG -> {
                shareParams.shareType = Platform.SHARE_IMAGE
                if (!TextUtils.isEmpty(imgShareBuilder.imagePath)) {
                    shareParams.imagePath = imgShareBuilder.imagePath
                    share(platform, shareParams)
                } else {
                    shareBitmap(imgShareBuilder.imageUrl, platform, shareParams)
                }
            }
            JShareUtil.ShareType.LINK -> {
                shareParams.title = linkShareBuilder.title
                shareParams.text = linkShareBuilder.text
                shareParams.shareType = Platform.SHARE_WEBPAGE
                shareParams.url = linkShareBuilder.url//必须
                if (!TextUtils.isEmpty(linkShareBuilder.imagePath)) {
                    shareParams.imagePath = linkShareBuilder.imagePath
                    share(platform, shareParams)
                } else {
                    shareBitmap(linkShareBuilder.imageUrl, platform, shareParams)
                }
            }
        }


    }

    fun share2Q(platform: String) {
        val shareParams = ShareParams()
        when (type) {
            JShareUtil.ShareType.TEXT -> {
                shareParams.shareType = Platform.SHARE_TEXT
                shareParams.text = textShareBuilder.text
            }
            JShareUtil.ShareType.IMG -> {
                shareParams.shareType = Platform.SHARE_IMAGE
                if (platform == QQ.Name) {
                    shareParams.imagePath = imgShareBuilder.imagePath
                } else if (platform == QZone.Name) {
                    if (TextUtils.isEmpty(imgShareBuilder.imageUrl)) {
                        shareParams.imagePath = imgShareBuilder.imagePath
                    } else {
                        shareParams.imageUrl = imgShareBuilder.imageUrl
                    }
                }

            }
            JShareUtil.ShareType.LINK -> {
                if (platform == QQ.Name) {
                    shareParams.title = subString(linkShareBuilder.title, 30)
                    shareParams.text = subString(linkShareBuilder.text, 40)
                }
                if (platform == QZone.Name) {
                    shareParams.title = subString(linkShareBuilder.title, 200)
                    shareParams.text = subString(linkShareBuilder.text, 600)
                }
                shareParams.shareType = Platform.SHARE_WEBPAGE
                shareParams.url = linkShareBuilder.url//必须
                if (TextUtils.isEmpty(linkShareBuilder.imageUrl)) {
                    shareParams.imagePath = linkShareBuilder.imagePath
                } else {
                    shareParams.imageUrl = linkShareBuilder.imageUrl
                }
            }

        }
        share(platform, shareParams)
    }


    fun share2Sina(platform: String) {
        val shareParams = ShareParams()
        when (type) {
            JShareUtil.ShareType.TEXT -> {
                shareParams.text = subString(textShareBuilder.text, 1999)
                shareParams.shareType = Platform.SHARE_TEXT
                shareParams.imageData = BitmapFactory.decodeResource(mActivity?.resources, R.mipmap.ic_share_weibo)
                share(SinaWeibo.Name, shareParams)
            }
            JShareUtil.ShareType.IMG -> {
                shareParams.text = subString(imgShareBuilder.text, 1999)
                shareParams.shareType = Platform.SHARE_IMAGE
                if (!TextUtils.isEmpty(imgShareBuilder.imagePath)) {
                    shareParams.imagePath = imgShareBuilder.imagePath
                    share(SinaWeibo.Name, shareParams)
                } else {
                    shareBitmap(imgShareBuilder.imageUrl, SinaWeibo.Name, shareParams)
                }
            }
            JShareUtil.ShareType.LINK -> {
                shareParams.text = subString(linkShareBuilder.text, 1999)
                shareParams.shareType = Platform.SHARE_WEBPAGE
                shareParams.url = linkShareBuilder.url
                if (!TextUtils.isEmpty(linkShareBuilder.imagePath)) {
                    shareParams.imagePath = linkShareBuilder.imagePath
                    share(platform, shareParams)
                } else {
                    shareBitmap(linkShareBuilder.imageUrl, platform, shareParams)
                }
            }
        }
    }

    fun share2Explorer() {
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(linkShareBuilder.url))
        mActivity?.startActivity(intent)
    }


    private fun subString(string: String?, length: Int): String? {
        return if (string != null && string.length > length) {
            string.substring(0, length)
        } else {
            string
        }

    }


    fun share(platForm: String, shareParams: ShareParams) {

        JShareInterface.share(platForm, shareParams, object : PlatActionListener {
            override fun onComplete(platform: Platform, i: Int, hashMap: HashMap<String, Any>) {
                println("分享成功")
            }

            override fun onError(platform: Platform, i: Int, i1: Int, throwable: Throwable) {
                println("分享失败" + throwable.toString())
                try {
                    if (platform.name == QQ.Name || platform.name == QZone.Name) {
                        mActivity?.toast("请安装QQ")
                    } else if (platform.name == Wechat.Name || platform.name == WechatFavorite.Name
                            || platform.name == WechatMoments.Name) {
                        mActivity?.toast("请安装微信")
                    } else if (platform.name == SinaWeibo.Name) {
                        mActivity?.toast("请安装微博")
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }

            }

            override fun onCancel(platform: Platform, i: Int) {
                println("分享取消")
            }
        })
    }


    fun shareBitmap(string: String?, platform: String, shareParams: ShareParams) {
        doAsyncResult {
            var bitmap: Bitmap
            try {
                val url = URL(string)
                val conn = url.openConnection() as HttpsURLConnection
                conn.requestMethod = "GET"
                conn.connectTimeout = 10 * 1000
                val inStream = conn.inputStream//通过输入流获取图片数据
                bitmap = BitmapFactory.decodeStream(inStream)
            } catch (e: Exception) {
                bitmap = BitmapFactory.decodeResource(MyApplication.instance.resources, R.mipmap.ic_launcher)
            }
            bitmap
            //compressBitmap(bitmap)
        }.get().let {
            shareParams.imageData = it
            share(platform, shareParams)
        }
    }


    /**
     * 当需要限制bitmap大小时,将图片压缩到指定大小之下
     *
     * @param b 原bitmap
     * @return 目标bitmap
     * @throws IOException
     */
    @Throws(IOException::class)
    private fun compressBitmap(b: Bitmap?): Bitmap? {
        var bitmap = b
        val topLimit = 30 * 1024
        if (bitmap != null) {
            val baos = ByteArrayOutputStream()
            var options = 100
            bitmap.compress(Bitmap.CompressFormat.JPEG, options, baos)
            Log.w("xxx", baos.toByteArray().size.toString() + "")
            while (baos.toByteArray().size > topLimit) {
                baos.reset()
                options -= 5
                bitmap.compress(Bitmap.CompressFormat.JPEG, options, baos)
                if (options == 0) {
                    break
                }
            }
            val isBm = ByteArrayInputStream(baos.toByteArray())// 把压缩后的数据baos存放到ByteArrayInputStream中
            bitmap = BitmapFactory.decodeStream(isBm, null, null)// 把ByteArrayInputStream数据生成
            isBm.close()
        }
        return bitmap
    }
}